src/Hosting/SystemdServices/src/SystemdNotifier.cs

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Text;
+using System.Net.Sockets;
+
+namespace Microsoft.Extensions.Hosting.SystemdServices
+{
+    public class SystemdNotifier : ISystemdNotifier
+    {
+        const string NOTIFY_SOCKET = "NOTIFY_SOCKET";
+
+        private readonly string _socketPath;
+
+        public SystemdNotifier() :
+            this(GetNotifySocketPath())
+        { }
+
+        // For testing
+        internal SystemdNotifier(string socketPath)
+        {
+            _socketPath = socketPath;
+        }
+
+        /// <inheritdoc />
+        public bool IsEnabled => _socketPath != null;
+
+        /// <inheritdoc />
+        public void Notify(ServiceState state, ServiceState[] states)
+        {
+            if (!IsEnabled)
+            {
+                return;
+            }
+
+            byte[] data;
```

Comment:
Can this buffer be rented from the `ArrayPool<byte>`? Would safe the allocation(s).

State is ASCII, so the rented size can be computed.
Same for the `states.Length != 0`-path. Size can be computed, rent the buffer, construct the MS from that buffer.

Reply:
Or stackalloced if the size will be small.

Reply:
I'll take a look at what is possible here based on the desired target tfm (netstandard2.0 or netstandard 2.1).

Reply:
+1 This method allocates too much. How often is it called in mainline uses cases?

Reply:
From earlier conversations with @glennc, I believe this is not a hot path at all. It's only used during an app state transition (starting, stopping, app-level errors, etc.).

`stackalloc`ing is reasonable but must only be done if the amount of space needed is finite and relatively small. Could always fall back to a rented buffer if the necessary space is too big.

Based on my understanding above, this doesn't feel like it needs super heavy optimization. One option is to use pre-defined byte arrays for the well-known values (see comment on ServiceState above).

Reply:
> I believe this is not a hot path at all.

That's correct. This gets called once on startup, and once on shutdown.

> One option is to use pre-defined byte arrays for the well-known values 

+1

Reply:
> once on startup, and once on shutdown.

Then I would stick with the current code, as it will be the most readable one.